pandas/core/internals/construction.py

```diff
@@ -240,18 +241,35 @@ def init_ndarray(values, index, columns, dtype: Optional[DtypeObj], copy: bool):
         if values.ndim == 2 and values.shape[0] != 1:
             # transpose and separate blocks
 
-            dvals_list = [maybe_infer_to_datetimelike(row) for row in values]
-            for n in range(len(dvals_list)):
-                if isinstance(dvals_list[n], np.ndarray):
-                    dvals_list[n] = dvals_list[n].reshape(1, -1)
+            dvals_list = (maybe_infer_to_datetimelike(row) for row in values)
 
             from pandas.core.internals.blocks import make_block
 
-            # TODO: What about re-joining object columns?
-            block_values = [
-                make_block(dvals_list[n], placement=[n], ndim=2)
-                for n in range(len(dvals_list))
-            ]
+            i = 0
+            block_values = []
+            for is_object, group in groupby(
+                dvals_list, lambda row: is_object_dtype(row.dtype)
+            ):
+                dval_group = list(group)
+                ei = i + len(dval_group)
+                if is_object:
+                    block_values.append(
+                        make_block(
+                            values[i:ei],
+                            placement=slice(i, ei),
+                            ndim=2,
+                        )
+                    )
+                else:
+                    block_values.extend(
+                        make_block(
```

Comment:
what happens here if we have two consecutive datetime64tz columns?

Reply:
```python
In[5]: block_values
Out[5]: 
[DatetimeTZBlock: slice(0, 1, 1), 1 x 1, dtype: datetime64[ns, UTC],
 DatetimeTZBlock: slice(1, 2, 1), 1 x 1, dtype: datetime64[ns, UTC],
 DatetimeTZBlock: slice(2, 3, 1), 1 x 1, dtype: datetime64[ns, UTC]]
```